org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java

```diff
@@ -59,15 +64,29 @@ public SynchronizationResult synchronize(IProgressMonitor monitor) {
 
     public SynchronizationResult synchronize(NewProjectHandler newProjectHandler, CancellationTokenSource tokenSource, IProgressMonitor monitor) {
         monitor = monitor != null ? monitor : new NullProgressMonitor();
-        SynchronizeOperation operation = new SynchronizeOperation(newProjectHandler);
+
+        SynchronizeOperation runningOperation = syncOperations.get(this);
+        if (runningOperation != null && (newProjectHandler == NewProjectHandler.NO_OP || Objects.equals(newProjectHandler, runningOperation.newProjectHandler))) {
+            return newSynchronizationResult(Status.OK_STATUS);
+        }
+
+        SynchronizeOperation operation = new SynchronizeOperation(this.gradleBuild, newProjectHandler);
+        syncOperations.put(this, operation);
```

Comment:
I think there's still a race condition here: if this method is called from multiple threads, both could get past the initial check and to this line. Instead, I'd call `putIfAbsent` and check if the return value is `null` in line 68.

Reply:
Good catch! I've adjusted the code.